### DISCUSSION TITLE
how to solve failed to load buitlin  ruff for mathod diagnosics pls chack your config

### DISCUSSION THREAD

**Discussion by sachinchaudhary1808** [nvimtools/none-ls.nvim#142] [Q&A] [answered]
### FAQ

- [X] I have checked the [FAQ](https://github.com/nvimtools/none-ls.nvim#faq) and it didn't resolve my problem.

### Issues

- [X] I have checked [existing issues](https://github.com/nvimtools/none-ls.nvim/issues?q=is%3Aissue) and there are no issues with the same problem.

### Neovim Version

0.10

### Dev Version?

- [X] I am using a stable Neovim release version, or if I am using a dev version of Neovim I have confirmed that my issue is reproducible on a stable version.

### Operating System

nixos

### Minimal Config

```Lua
local null_ls = require("null-ls")

null_ls.setup({
	sources = {
		null_ls.builtins.formatting.stylua,
		null_ls.builtins.completion.spell,
		-- require("none-ls.diagnostics.eslint"), -- requires none-ls-extras.nvim
		null_ls.builtins.diagnostics.mypy,
		null_ls.builtins.diagnostics.ruff,
	},
})

vim.keymap.set("n", "<leader>=", vim.lsp.buf.format, { desc = "formattes code using none-ls" })
```

### Steps to Reproduce

use the code above

### Reproducibility Check

- [X] I confirm that my minimal config is based on the `minimal_init.lua` template and that my issue is reproducible by running `nvim --clean -u minimal_init.lua` and following the steps above.

### Expected Behavior

so like i use nix as pkg manager i searched about this issue on web and they said i need some exra thing which is not in nixpkgs

### Actual Behavior

above

### Debug Log

i think we don't need that 

### Help

No

### Implementation Help

add that plugin into nixpkgs or help me 

### Requirements

- [X] I have read and followed the instructions above and understand that my issue will be closed if I did not provide the required information.

**Comment by sachinchaudhary1808** [comment]
i'm so stupid sorry

**Reply by gsavarela** [reply]
@sachinchaudhary1808 I have been having the same issue with ruff:

> [null-ls] failed to load builtin ruff for method diagnostics; please check your config

How did you fix it?

**Reply by sachinchaudhary1808** [reply]
> @sachinchaudhary1808 I have been having the same issue with ruff:
> 
> > [null-ls] failed to load builtin ruff for method diagnostics; please check your config
> 
> How did you fix it?

i just use pylint for now

**Reply by gsavarela** [reply]
Hi thanks for the reply -- so I guess we should re open the issue since this problem is not solved. Right?
(Pytlint can be 120x slower than ruff).

**Comment by sachinchaudhary1808** [comment]
how can i get formate on save i'm pretty new and i don't have any idea

**Reply by gsavarela** [reply]
I have this on my null-ls.lua config file:

```

local autogroup = vim.api.nvim_create_augroup("LspFormatting", {})
local null_ls =   require('null-ls')

local opts = {
  sources = {
    null_ls.builtins.formatting.black,
    null_ls.builtins.diagnostics.mypy,
    null_ls.builtins.diagnostics.ruff,
  },
  on_attach = function(client, bufnr)
    if client.supports_method("textDocument/formatting") then
      vim.api.nvim_clear_autocmds({
        group = autogroup,
        buffer = bufnr,
      })
      vim.api.nvim_create_autocmd("BufWritePre", {
        group = autogroup,
        buffer = bufnr,
        callback = function()
          vim.lsp.buf.format({bufnr = bufnr})
        end,

      })
    end
  end,
}
return opts
```

For thorough instructions refer to this [tutorial](https://www.youtube.com/watch?v=4BnVeOUeZxc&t=550s)

**Reply by sachinchaudhary1808** [reply]
> I have this on my null-ls.lua config file:
> 
> ```
> 
> local autogroup = vim.api.nvim_create_augroup("LspFormatting", {})
> local null_ls =   require('null-ls')
> 
> local opts = {
>   sources = {
>     null_ls.builtins.formatting.black,
>     null_ls.builtins.diagnostics.mypy,
>     null_ls.builtins.diagnostics.ruff,
>   },
>   on_attach = function(client, bufnr)
>     if client.supports_method("textDocument/formatting") then
>       vim.api.nvim_clear_autocmds({
>         group = autogroup,
>         buffer = bufnr,
>       })
>       vim.api.nvim_create_autocmd("BufWritePre", {
>         group = autogroup,
>         buffer = bufnr,
>         callback = function()
>           vim.lsp.buf.format({bufnr = bufnr})
>         end,
> 
>       })
>     end
>   end,
> }
> return opts
> ```
> 
> For thorough instructions refer to this [tutorial](https://www.youtube.com/watch?v=4BnVeOUeZxc&t=550s)

thanks

**Comment by mochaaP** [comment, marked as answer]
See #81

**Reply by mochaaP** [reply]
tldr: use ruff's native lsp: `ruff server --preview`.
<https://github.com/neovim/nvim-lspconfig/blob/master/doc/server_configurations.md#ruff>

**Reply by mochaaP** [reply]
@sachinchaudhary1808 @gsavarela

**Reply by gsavarela** [reply]
Thanks @mochaaP @sachinchaudhary1808 -- Issue was non-issue. So solved.
